beacon-chain/rpc/validator/proposer.go

```diff
@@ -194,6 +206,128 @@ func (vs *Server) eth1Data(ctx context.Context, slot uint64) (*ethpb.Eth1Data, e
+// TODO: "Genesis" period and previous period
+// TODO: Feature flag
+func (vs *Server) eth1DataMajorityVote(ctx context.Context, slot uint64) (*ethpb.Eth1Data, error) {
+	ctx, cancel := context.WithTimeout(ctx, eth1dataTimeout)
+	defer cancel()
+
+	if vs.MockEth1Votes {
+		return vs.mockETH1DataVote(ctx, slot)
+	}
+	if !vs.Eth1InfoFetcher.IsConnectedToETH1() {
+		return vs.randomETH1DataVote(ctx)
+	}
+	eth1DataNotification = false
+
+	headOfVotingPeriod :=
+		slot%(params.BeaconConfig().EpochsPerEth1VotingPeriod*params.BeaconConfig().SlotsPerEpoch) <
+			params.BeaconConfig().HeadOfVotingPeriodLength
+	if headOfVotingPeriod {
+		// TODO: Is it OK that `context.WithTimeout` will be invoked again?
+		return vs.eth1Data(ctx, slot)
+	}
+
+	slotsPerVotingPeriod := params.BeaconConfig().EpochsPerEth1VotingPeriod * params.BeaconConfig().SlotsPerEpoch
+	genesisTime, _ := vs.Eth1InfoFetcher.Eth2GenesisPowchainInfo()
+	currentPeriodVotingStartTime := genesisTime +
+		(slot-(slot%slotsPerVotingPeriod))*params.BeaconConfig().SecondsPerSlot
+	previousPeriodVotingStartTime := currentPeriodVotingStartTime -
+		slotsPerVotingPeriod*params.BeaconConfig().SecondsPerSlot
+
+	currentPeriodBlockNumber, err := vs.Eth1BlockFetcher.BlockNumberByTimestamp(ctx, currentPeriodVotingStartTime)
+	if err != nil {
+		log.WithError(err).Error("Failed to get block number for current voting period")
+		return vs.randomETH1DataVote(ctx)
+	}
+	previousPeriodBlockNumber, err := vs.Eth1BlockFetcher.BlockNumberByTimestamp(ctx, previousPeriodVotingStartTime)
+	if err != nil {
+		log.WithError(err).Error("Failed to get block number for previous voting period")
+		return vs.randomETH1DataVote(ctx)
+	}
+
+	headState, err := vs.HeadFetcher.HeadState(ctx)
+	if err != nil {
+		return nil, status.Error(codes.Internal, "Could not get head state")
+	}
+	if len(headState.Eth1DataVotes()) == 0 {
+		eth1Data, err := vs.defaultEth1DataResponse(ctx, currentPeriodBlockNumber)
+		if err != nil {
+			log.WithError(err).Error("Failed to get eth1 data from current period block number")
+			return vs.randomETH1DataVote(ctx)
+		}
+		return eth1Data, nil
+	}
+
+	eth1FollowDistance := int64(params.BeaconConfig().Eth1FollowDistance)
+	lastValidBlockNumber := big.NewInt(0).Sub(currentPeriodBlockNumber, big.NewInt(eth1FollowDistance))
+	firstValidBlockNumber := big.NewInt(0).Sub(previousPeriodBlockNumber, big.NewInt(eth1FollowDistance))
+
+	var inRangeVotes []eth1DataSingleVote
+	for _, eth1Data := range headState.Eth1DataVotes() {
+		ok, height, err := vs.BlockFetcher.BlockExists(ctx, bytesutil.ToBytes32(eth1Data.BlockHash))
+		if err != nil {
+			log.WithError(err).Warning("Could not fetch eth1data height for received eth1data vote")
+		}
+		// TODO: Both inclusive?
+		if ok && lastValidBlockNumber.Cmp(height) > -1 && firstValidBlockNumber.Cmp(height) < 1 {
+			inRangeVotes = append(inRangeVotes, eth1DataSingleVote{eth1Data: *eth1Data, blockHeight: height})
+		}
```

Comment:
The linked document describes the algorithm in such a way that both the upper and lower bound blocks are included in the possible result set. Because of this the same block can be potentially voted for in two succeeding periods (the upper bound for the previous period becomes the lower bound for the next period). Are we OK with this or should one of the bounds be exclusive?